File: WINGS_AI_Bootcamp/Scala 3/numToWords.scala
@main def numberToWords(n: Long): Unit = {
  val ones = Array("", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine",
    "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen",
    "Seventeen", "Eighteen", "Nineteen")
  val tens = Array("", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety")

  def twoDigits(num: Int): String =
    if (num < 20) ones(num)
    else tens(num / 10) + (if (num % 10 != 0) " " + ones(num % 10) else "")

  def threeDigits(num: Int): String =
    if (num == 0) ""
    else if (num < 100) twoDigits(num)
    else ones(num / 100) + " Hundred" + (if (num % 100 != 0) " " + twoDigits(num % 100) else "")

  def convertIndian(num: Long): String = {
    if (num == 0) return "Zero"
    val parts = Array(
      (10000000, "Crore"),
      (100000, "Lakh"),
      (1000, "Thousand"),
      (100, "Hundred")
    )

    var n = num
    val sb = new StringBuilder

    for ((divisor, label) <- parts) {
      val part = (n / divisor).toInt
      if (part > 0) {
        if (divisor == 100) sb.append(ones(part)).append(" ").append(label)
        else sb.append(threeDigits(part)).append(" ").append(label)
        sb.append(" ")
        n %= divisor
      }
    }


    if (n > 0) sb.append(twoDigits(n.toInt))
    sb.toString.trim
  }

  println(convertIndian(n))
}

File: WINGS_AI_Bootcamp/Elixir/numToWords.exs
defmodule NumberToWords do
  @ones ~w(zero one two three four five six seven eight nine ten eleven twelve thirteen fourteen fifteen sixteen seventeen eighteen nineteen)
  @tens ~w(zero ten twenty thirty forty fifty sixty seventy eighty ninety)

  @indian_units [
    {10000000, "crore"},
    {100000, "lakh"},
    {1000, "thousand"},
    {100, "hundred"}
  ]

  @western_units [
    {1_000_000_000, "billion"},
    {1_000_000, "million"},
    {1000, "thousand"},
    {100, "hundred"}
  ]

  def convert(n, :indian) when n < 100_00_00_000, do: to_words(n, @indian_units)
  def convert(n, :western) when n < 1_000_000_000, do: to_words(n, @western_units)

  defp to_words(0, _), do: "zero"
  defp to_words(n, units) do
    build_words(n, units) |> Enum.join(" ") |> String.trim()
  end

  defp build_words(n, units) do
    Enum.reduce(units, {n, []}, fn {val, label}, {num, acc} ->
      if num >= val do
        div_part = div(num, val)
        rem_part = rem(num, val)
        words = build_words(div_part, @indian_units)
        {rem_part, acc ++ words ++ [label]}
      else
        {num, acc}
      end
    end)
    |> then(fn {num, acc} ->
      acc ++ two_digit_words(num)
    end)
  end

  defp two_digit_words(0), do: []
  defp two_digit_words(n) when n < 20, do: [Enum.at(@ones, n)]
  defp two_digit_words(n) do
    tens = div(n, 10)
    ones = rem(n, 10)
    [Enum.at(@tens, tens)] ++ if ones > 0, do: [Enum.at(@ones, ones)], else: []
  end
end

# Example usage
IO.puts NumberToWords.convert(12345678, :indian)
IO.puts NumberToWords.convert(12345678, :western)


